radix-engine/src/model/resource.rs

```diff
@@ -0,0 +1,282 @@
+    /// The resource type
+    resource_type: ResourceType,
+    /// The contained resource
+    resource: Resource,
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum ResourceContainerId {
+    /// For named bucket
+    Bucket(BucketId),
+    /// For vault
+    Vault(VaultId),
+
+    /// For the specific resource on the n-th worktop
+    Worktop {
+        depth: u32,
+        resource_def_id: ResourceDefId,
+    },
+}
+
+#[derive(Debug, Clone)]
+pub struct Proof {
+    /// The resource definition id
+    resource_def_id: ResourceDefId,
+    /// The resource type
+    resource_type: ResourceType,
+    /// Restricted proof can't be moved down along the call stack (growing down).
+    restricted: bool,
+    /// The total amount locked
+    total_amount: ResourceAmount,
+    /// The sub-amounts (to be extended)
+    #[allow(dead_code)]
+    amounts: HashMap<ResourceContainerId, (Rc<ResourceContainer>, ResourceAmount)>,
+}
+
+impl Resource {
+    pub fn fungible(amount: Decimal) -> Self {
+        Resource::Fungible {
+            locked_amounts: BTreeMap::new(),
+            liquid_amount: amount,
+        }
+    }
+
+    pub fn non_fungible(ids: BTreeSet<NonFungibleId>) -> Self {
+        Resource::NonFungible {
+            locked_ids: HashMap::new(),
+            liquid_ids: ids.clone(),
+        }
+    }
+}
+
+impl ResourceAmount {
+    pub fn quantity(&self) -> Decimal {
+        match self {
+            ResourceAmount::Fungible { amount } => *amount,
+            ResourceAmount::NonFungible { ids } => ids.len().into(),
+        }
+    }
+
+    pub fn non_fungible_ids(&self) -> Result<BTreeSet<NonFungibleId>, ResourceError> {
+        match self {
+            ResourceAmount::Fungible { .. } => Err(ResourceError::UnsupportedOperation),
+            ResourceAmount::NonFungible { ids } => Ok(ids.clone()),
+        }
+    }
+}
+
+impl ResourceContainer {
+    pub fn new(
+        resource_def_id: ResourceDefId,
+        resource_type: ResourceType,
+        resource: Resource,
+    ) -> Self {
+        Self {
+            resource_def_id,
+            resource_type,
+            resource,
+        }
+    }
+
+    pub fn put(&mut self, other: Self) -> Result<(), ResourceError> {
+        // check resource address
+        if self.resource_def_id != other.resource_def_id {
+            return Err(ResourceError::ResourceAddressNotMatching);
+        }
+
+        // assumption: owned bucket should not be locked
+        assert!(!other.is_locked());
+
+        // add the other bucket into liquid pool
+        match (&mut self.resource, other.liquid_amount()) {
+            (Resource::Fungible { liquid_amount, .. }, ResourceAmount::Fungible { amount }) => {
+                *liquid_amount = *liquid_amount + amount;
+            }
+            (Resource::NonFungible { liquid_ids, .. }, ResourceAmount::NonFungible { ids }) => {
+                liquid_ids.extend(ids);
+            }
+            _ => panic!("Resource type should match!"),
```

Comment:
This is awesome matching logic :)

Reply:
Haha, I've been struggling to find an alternative...